controllers/teleport_controller/teleport_controller.cpp

```diff
@@ -0,0 +1,138 @@
+
+        // Output current location
+        std::cout << "Location: " << std::endl;
+        std::cout << "X: " << target_translation_vec[0];
+        std::cout << " Y: " << target_translation_vec[1];
+        std::cout << " Z: " << target_translation_vec[2] << std::endl;
+
+        // Prepare new location
+
+        /* 0, 0, 0 is centre of the playing field.
+         *     X ranges from -5.4 - 5.4, the positive value on the left hand side when looking
+         *     from red
+         *     Y ranges from -8 - 8, the positive value on the red goal side
+         *     Z should be set to 3.2 to be on the ground
+         *     If the robot teleports into an existing object it may run into issues for that
+         *     image only, after resetPhysics is should return to a regular state
+         */
+
+        double newPos[3];
+        newPos[0] = 5.4 - (rand() % 1080) / 100;  // target_translation_vec[0] + 1;
+        newPos[1] = 8 - (rand() % 1600) / 100;    // target_translation_vec[1] + 0;
+        newPos[2] = 0.32;                         // target_translation_vec[2] + 0;
+
+        // Set new location
+        target_translation_field->setSFVec3f(newPos);
+
+        // Grab the current rotation field of the robot to modify
+        Field* target_rotation_field = target->getField("rotation");
+        // Convert the field to a vector to output to console
+        const double* target_rotation_vec = target_rotation_field->getSFRotation();
+
+        // Loop 4 times, one for each direction to face
+        for (int i = 0; i < rotationsSize; i++) {
+            // Output current rotation
+            std::cout << "Rotation: " << std::endl;
+            std::cout << "X: " << target_rotation_vec[0];
+            std::cout << " Y: " << target_rotation_vec[1];
+            std::cout << " Z: " << target_rotation_vec[2];
+            std::cout << " α: " << target_rotation_vec[3] << std::endl;
+
+            /* Prepare new rotation. These are saved in rotations matrix as the axis-angle
+               calculation is rough to calculate on the fly*/
+
+            double newRot[4];
+            newRot[0] = rotations[i][0];
+            newRot[1] = rotations[i][1];
+            newRot[2] = rotations[i][2];
+            newRot[3] = rotations[i][3];
+
+            // Set new rotations
+            target_rotation_field->setSFRotation(newRot);
+
+            // This line is very important. The robot blows itself up after a 1:05 minutes otherwise
+            // Teleporting is strenuous work apparently
+            target->resetPhysics();
+
+            /* Make robot stand up and look around. This preset is slow and I'm not super
+             *     happy with it but it does the job for now. Below is a work in progress to
+             *     manually move the robot motors to where we want them
+             */
+            mMotionManager->playPage(54);
+            // Other presets, comment out as desired
```

Comment:
should these be commented out? this comment makes it seem like they should before this is merged??

Reply:
Not knowing the exact way that the program was going to be used I have a couple of options in there that someone can comment out as they wish. They are meant to be uncommented but I wanted it to be clear that removing them wouldn't break the program if someone wanted to change the preset motions that the robot was performing

Reply:
That makes some sense to me. I think that some of that info could be added to the comment though

Reply:
or I suppose, ideally this would be a config value in a config file, so you know that this is something which changes how the thing runs, and where to change it. The description you just gave would be condensed to stuff relevant to someone changing the config file later.

Reply:
Magic numbers in general are bad ([see this wiki page about "unnamed numerical constants" for more info on what a magic number is](https://en.wikipedia.org/wiki/Magic_number_(programming))). Magic numbers in config files with descriptions/comments/explanations (which I guess means that they wouldn't be magic numbers any more by definition lol) are much better.

Reply:
Are you proposing having it read from a config file which you can edit? Or just have a txt that explains what's going on?
And yes silly oversight on the numbers thing, I had the reference open in front of me so I knew exactly what was going on 😆

Reply:
Basically if it is a constant and has meaning, stick it in a variable and give the variable a name

Reply:
Would this be something I should use a #define for or just a regular constant is okay?

Reply:
You should ever really be using #define in c++, there are very few situations where it's a good idea. The main reason being they don't have type information and are just doing text replacement so you can get into some weird spots.

In c++ we use `constexpr` to define constants with a type like

```cpp
constexpr int ULTIMATE_ANSWER = 42;
```

Reply:
Yeah okay sweet, ty